paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/operator/BatchReadSource.java

```diff
@@ -0,0 +1,140 @@
+ * parallelism greater than one.
+ */
+public class BatchReadSource extends AbstractNonCoordinatedSource<Split> {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(BatchReadSource.class);
+
+    private final ReadBuilder readBuilder;
+
+    private final String tableName;
+
+    public BatchReadSource(ReadBuilder readBuilder) {
+        this.readBuilder = readBuilder;
+        this.tableName = readBuilder.tableName();
+    }
+
+    @Override
+    public Boundedness getBoundedness() {
+        return Boundedness.BOUNDED;
+    }
+
+    @Override
+    public SourceReader<Split, SimpleSourceSplit> createReader(
+            SourceReaderContext sourceReaderContext) throws Exception {
+        return new Reader();
+    }
+
+    private class Reader extends AbstractNonCoordinatedSourceReader<Split> {
+
+        private final TableScan scan = readBuilder.newScan();
+
+        @Override
+        public InputStatus pollNext(ReaderOutput<Split> readerOutput) throws Exception {
+            List<Split> splits = scan.plan().splits();
+
+            if (splits.size() == 0) {
+                LOG.info("In BatchSource the table {} plan split is empty.", tableName);
+            } else {
+                DataSplit dataSplit = (DataSplit) splits.get(0);
+                LOG.info(
+                        "Read the snapshot-{} of table {} in BatchSource.",
+                        dataSplit.snapshotId(),
+                        tableName);
+            }
+
+            splits.forEach(readerOutput::collect);
+
+            return InputStatus.END_OF_INPUT;
+        }
+    }
+
+    public static DataStream<RowData> buildSource(
+            StreamExecutionEnvironment env,
+            String name,
+            TypeInformation<RowData> typeInfo,
+            ReadBuilder readBuilder,
+            boolean shuffleBucketWithPartition,
+            boolean unawareBucket,
+            NestedProjectedRowData nestedProjectedRowData,
+            Long limit) {
+        SingleOutputStreamOperator<Split> singleOutputStreamOperator =
+                env.fromSource(
+                                new BatchReadSource(readBuilder),
+                                WatermarkStrategy.noWatermarks(),
+                                name + "-BatchRead",
+                                new JavaTypeInfo<>(Split.class))
+                        .forceNonParallel();
+
+        DataStream<Split> sourceDataStream =
+                unawareBucket
+                        ? shuffleUnawareBucket(singleOutputStreamOperator)
+                        : shuffleNonUnawareBucket(
+                                singleOutputStreamOperator, shuffleBucketWithPartition);
+
+        return sourceDataStream.transform(
+                name + "-Reader",
+                typeInfo,
+                new ReadOperator(readBuilder::newRead, nestedProjectedRowData, limit));
```

Comment:
Unlike streaming mode, in batch mode Flink would enable auto parallelism inference by default. Given that the parallelism of downstream operators would be inferred according to `execution.batch.adaptive.auto-parallelism.avg-data-volume-per-task`, and the plan splits transmitted here are usually of very small size, I'm a little concerned that the parallelism of readOperator will be inferred into a small value, and users have to manually set the parallelism of this operator to the correct range, which obeys the convenient brought by auto parallelism inference. What do you think?

Reply:
Thanks very much, you are right.
I have added some description in the option's description.